modules/tinymce/src/core/main/ts/keyboard/Nbsps.ts

```diff
@@ -149,22 +162,35 @@ const normalizeNbspInMiddleOfTextNode = (node: Text): boolean => {
   }
 };
 
-const normalizeNbspAtEnd = (root: SugarElement<Node>, node: Text): boolean => {
+const normalizeNbspAtEnd = (root: SugarElement<Node>, node: Text, makeNbsp: boolean): boolean => {
   const text = node.data;
   const lastPos = CaretPosition(node, text.length - 1);
-  if (isNbspAt(text, text.length - 1) && !needsToBeNbsp(root, lastPos)) {
+  if (!makeNbsp && isNbspAt(text, text.length - 1) && !needsToBeNbsp(root, lastPos)) {
     node.data = text.slice(0, -1) + ' ';
     return true;
+  } else if (makeNbsp && isWhiteSpaceAt(text, text.length - 1) && needsToBeNbspRight(root, lastPos)) {
+    node.data = text.slice(0, -1) + Unicode.nbsp;
+    return true;
   } else {
     return false;
   }
 };
 
-const normalizeNbsps = (root: SugarElement<Node>, pos: CaretPosition): Optional<CaretPosition> => Optional.some(pos).filter(hasNbsp).bind((pos) => {
-  const container = pos.container() as Text;
-  const normalized = normalizeNbspAtStart(root, container) || normalizeNbspInMiddleOfTextNode(container) || normalizeNbspAtEnd(root, container);
-  return normalized ? Optional.some(pos) : Optional.none();
-});
+const normalizeNbsps = (root: SugarElement<Node>, pos: CaretPosition): Optional<CaretPosition> => {
+  if (hasNbsp(pos)) {
+    const container = pos.container() as Text;
+    const normalized = normalizeNbspAtStart(root, container, false) || normalizeNbspInMiddleOfTextNode(container) || normalizeNbspAtEnd(root, container, false);
+    return Optionals.someIf(normalized, pos);
+  } else {
+    if (needsToBeNbsp(root, pos) && NodeType.isText(pos.container())) {
+      const container = pos.container() as Text;
+      const normalized = normalizeNbspAtStart(root, container, true) || normalizeNbspAtEnd(root, container, true);
+      return Optionals.someIf(normalized, pos);
+    } else {
+      return Optional.none();
+    }
+  }
```

Comment:
You've got nested ifs here when you don't need them, nor does it really make sense to have them nested.
```suggestion
  } else if (needsToBeNbsp(root, pos) && NodeType.isText(pos.container())) {
    const container = pos.container() as Text;
    const normalized = normalizeNbspAtStart(root, container, true) || normalizeNbspAtEnd(root, container, true);
    return Optionals.someIf(normalized, pos);
  } else {
    return Optional.none();
  }
```

Reply:
True, I shorted it down without thinking, that's on me.